tutorials/haskell/ghc-optimization-and-fusion/tutorial.md

```diff
@@ -0,0 +1,1945 @@
+performs:
+
+```
+benchmarking nofusion0
+time                 155.4 ms   (146.4 ms .. 162.4 ms)
+                     0.996 R²   (0.980 R² .. 1.000 R²)
+mean                 155.1 ms   (151.3 ms .. 159.0 ms)
+std dev              5.522 ms   (3.154 ms .. 7.537 ms)
+```
+
+This is the result with `[0..1000000]` passed as argument to `nofusion0`.
+
+With `weigh` (a relatively new library that allows to find out memory
+consumption of your code) I'm getting the following:
+
+```
+Case                  Bytes  GCs  Check
+nofusion0       249,259,656  448  OK
+```
+
+In a lazy language like Haskell laziness just changes when parts of
+intermediate lists are allocated, but they still must be allocated because
+that's what the next step in the “pipe” takes as input, and that's the
+overhead we want to reduce with fusion.
+
+Can we do better if we rewrite everything as a single function that sums and
+multiplies in one pass? Not so sexy, but let's give it a try to see what
+sort of power we're missing with our “elegant” composition of list
+processing functions:
+
+```haskell
+manuallyFused :: [Int] -> Int
+manuallyFused []     = 0
+manuallyFused (x:xs) = x * x + manuallyFused xs
+```
+
+Let's benchmark it:
+
+```
+benchmarking manuallyFused
+time                 17.10 ms   (16.71 ms .. 17.54 ms)
+                     0.996 R²   (0.992 R² .. 0.998 R²)
+mean                 17.18 ms   (16.87 ms .. 17.62 ms)
+std dev              932.8 μs   (673.7 μs .. 1.453 ms)
+
+Case                 Bytes  GCs  Check
+manuallyFused   96,646,160  153  OK
+```
+
+The improvement is dramatic. We just manually “fused” the two functions and
+produced code that runs faster, consumes less memory, and does the same
+thing. But how can we give up on composability and elegance — main merits of
+functional programming? No way!
+
+What we would like to achieve is the following:
+
+1. Ability to write beautiful, composable programs.
+2. Avoid allocating intermediate results where possible, because it sucks.
+
+The point 2 can be (and has been) addressed differently:
+
+1. We can build our vocabulary of little “primitive” operations (that we use
+   as building blocks in our programs) in such a way that they do not ever
+   produce results immediately. So when such primitives are combined, they
+   produce another (wrapped) function that does not produce a result
+   immediately either. To get “real” result, we need yet another function
+   that can “run” the composite action we've constructed. This is also
+   fusion and this is how the `repa` package works for example.
+
+2. We want to have our cake and eat it too. We can expose familiar interface
+   where every “primitive” produces a result immediately, but we also can
+   add rewrite rules that will (hopefully) make GHC rewrite things in such a
+   way that in the end the compiler gets one tight loop without intermediate
+   allocations.
+
+I must say that I like the approach 1 more because it's more explicit and
```

Comment:
Why to remove "the" here? Also: https://english.stackexchange.com/questions/4988/like-something-more-or-like-something-better.